### ISSUE TITLE
Assumptions about grep

### ISSUE THREAD

**Issue by radicimo** [speed47/spectre-meltdown-checker#95]
Variant 3 will return a false positive on some ARM-based Android systems that implement busybox, because /system/bin/grep is rather brain dead and cannot handle the following:

`grep -qi 'CPU implementer : 0x41' /proc/cpuinfo`

However, using the grep that is installed as part of busybox does function properly.

The script makes assumptions about grep that may not hold true on all Linux variants. Please consider implementing some way to override the grep command in the script to allow the use of 'busybox grep' instead. It is possible to do a search and replace, but some of the variables have grep in their name, and this must be done carefully.

**Comment by speed47** [comment]
Thanks for your report.
I try to make this script as compatible as possible, which is why it relies on sh instead of bash, and I avoid using complicated commands (such as grep -P that's not available everywhere).
I'll add a param to ask the use of busybox. Can I have the output of your `busybox grep --help` ? I'll see which params are supported this way

**Comment by radicimo** [comment]
Usage: grep [-HhnlLoqvsriwFE] [-m N] [-A/B/C N] PATTERN/-e PATTERN.../-f FILE [FILE]...

Search for PATTERN in FILEs (or stdin)

        -H      Add 'filename:' prefix
        -h      Do not add 'filename:' prefix
        -n      Add 'line_no:' prefix
        -l      Show only names of files that match
        -L      Show only names of files that don't match
        -c      Show only count of matching lines
        -o      Show only the matching part of line
        -q      Quiet. Return 0 if PATTERN is found, 1 otherwise
        -v      Select non-matching lines
        -s      Suppress open and read errors
        -r      Recurse
        -i      Ignore case
        -w      Match whole words only
        -x      Match whole lines only
        -F      PATTERN is a literal (not regexp)
        -E      PATTERN is an extended regexp
        -m N    Match up to N times per file
        -A N    Print N lines of trailing context
        -B N    Print N lines of leading context
        -C N    Same as '-A N -B N'
        -e PTRN Pattern to match
        -f FILE Read pattern from file

**Comment by radicimo** [comment]
All I had to do was change 'grep' to 'busybox grep' and it handled the flags you are using and the pattern matching. The /system/bin/grep appears to be failing with the PATTERN matching:

```
# /system/bin/grep 'CPU implementer\s*:\s*0x41' /proc/cpuinfo                                                              
1
```

```
# busybox grep 'CPU implementer\s*:\s*0x41' /proc/cpuinfo                                                                  
CPU implementer : 0x41
CPU implementer : 0x41
```

```
# /system/bin/grep '0x41' /proc/cpuinfo                                                                                      
CPU implementer : 0x41
CPU implementer : 0x41
```

**Comment by speed47** [comment]
I've made modifications that you can test on the `arm_display` branch.
It should work without using busybox: I've just changed the `\s` by `[[:space:]]` that appear to work with simple versions of grep.
Could you try the `arm_display` branch?

**Comment by radicimo** [comment]
The modifications in arm_display branch (replacing the \s by [[:space:]]) do work on the system under test. Thank you for the fixes.
